.github/workflows/build.yaml

```diff
@@ -1,63 +1,125 @@
-name: push
+name: Build
 on:
   push:
     branches:
-      - master
       - main
     tags:
       - v*
 
+env:
+  DEFAULT_TAG_VERSION: "latest"
+  BUILD_PLATFORMS: "linux/amd64,linux/arm64"
+  IMAGE_REGISTRY: "ghcr.io"
+
 jobs:
+  setup:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Setup Workflow Variables
+        id: set-variables
+        run: |
+          echo "REPOSITORY_NAME=$(basename $GITHUB_REPOSITORY)" >> $GITHUB_OUTPUT
+          echo "IMAGE_REPOSITORY=${{ env.IMAGE_REGISTRY}}/$GITHUB_REPOSITORY" >> $GITHUB_OUTPUT
+
+          # Set versions based on presence of tag
+          if [[ "${{ github.ref }}" =~ ^refs/tags/ ]]; then
+            echo "IMAGE_TAG=$(echo ${GITHUB_REF/refs\/tags\//})" >> $GITHUB_OUTPUT
+          else
+            echo "IMAGE_TAG=${DEFAULT_TAG_VERSION}" >> $GITHUB_OUTPUT
+          fi
+
+          # Create Distribution Matrix
+          echo "DIST_MATRIX=$(echo -n "$BUILD_PLATFORMS" | jq -csR '. | split(",")')" >> $GITHUB_OUTPUT
+          # Create Image Tags
+          echo "IMAGE_PLATFORM_TAGS=$(echo $BUILD_PLATFORMS | sed  -e 's/,/ /g' -e 's/\//-/g')" >> $GITHUB_OUTPUT
+
+    outputs:
+      repository_name: ${{ steps.set-variables.outputs.REPOSITORY_NAME }}
+      image_repository: ${{ steps.set-variables.outputs.IMAGE_REPOSITORY }}
+      image_tag: ${{ steps.set-variables.outputs.IMAGE_TAG}}
+      dist_matrix: ${{ steps.set-variables.outputs.DIST_MATRIX }}
+      image_platform_tags: ${{ steps.set-variables.outputs.IMAGE_PLATFORM_TAGS }}
+
+
   build:
     runs-on: ubuntu-latest
+    needs: ["setup"]
     strategy:
       matrix:
-        platform: ['linux/amd64', 'linux/arm64']
+        platform: ${{ fromJson(needs.setup.outputs.dist_matrix) }}
     steps:
-      - name: set repo name
-        shell: bash
+      - name: Setup Job Variables
+        id: set-build-variables
         run: |
-          echo "REPOSITORY_NAME=$(basename $GITHUB_REPOSITORY)" >> $GITHUB_ENV
+          echo "PLATFORM_OS=$(echo ${{ matrix.platform }} |  cut -d/ -f1)" >> $GITHUB_OUTPUT
+          echo "PLATFORM_ARCH=$(echo ${{ matrix.platform }} |  cut -d/ -f2)" >> $GITHUB_OUTPUT
+
       - name: checkout repo
-        uses: actions/checkout@v2
+        uses: actions/checkout@755da8c3cf115ac066823e79a1e1788f8940201b # pin@v3
 
       - name: Set up QEMU
-        uses: docker/setup-qemu-action@v1
+        uses: docker/setup-qemu-action@e81a89b1732b9c48d79cd809d8d81d79c4647a18 #pin@v2.1.0
 
       - name: Set up Docker Buildx
-        uses: docker/setup-buildx-action@v1
+        uses: docker/setup-buildx-action@8c0edbc76e98fa90f69d9a2c020dcb50019dc325 #pin@v2.2.1
 
       - name: Docker Login
-        uses: docker/login-action@v1.12.0
+        uses: docker/login-action@f4ef78c080cd8ba55a85445d5b36e214a81df20a #pin@v2.1.0
         with:
-          registry: "ghcr.io"
+          registry: ${{ env.IMAGE_REGISTRY }}
```

Comment:
Maybe a nit. Does this login action make sense for any other registry than ghcr? I think the username and password would be different in such case, wouldn't it?

Reply:
Yes it would be as you would need to provide those as options.  Let's add this as a discussion point in an issue